ffi/zipwriter.lua

```diff
@@ -50,153 +16,45 @@ function ZipWriter:new()
-        if lfh._no_compression then
-            zh:write(numberToByteString(0, 2)) -- compression method: store
-        else
-            zh:write(numberToByteString(8, 2)) -- compression method: deflate
-        end
-        zh:write(numberToByteString(lfh.last_mod_file_time, 2))
-        zh:write(numberToByteString(lfh.last_mod_file_date, 2))
-        zh:write(numberToByteString(lfh.crc32, 4))
-        zh:write(numberToByteString(lfh.compressed_size, 4))
-        zh:write(numberToByteString(lfh.uncompressed_size, 4))
-        zh:write(numberToByteString(lfh.file_name_length, 2))
-        zh:write(numberToByteString(lfh.extra_field_length, 2))
-        zh:write(numberToByteString(0, 2)) -- file comment length
-        zh:write(numberToByteString(0, 2)) -- disk number start
-        zh:write(numberToByteString(0, 2)) -- internal file attributes
-        zh:write(numberToByteString(lfh.external_attr, 4)) -- external file attributes
-        zh:write(numberToByteString(lfh.offset, 4)) -- relative offset of local header
-        zh:write(lfh.file_name)
-        size_of_central_directory = size_of_central_directory + 46 + lfh.file_name_length
-    end
-    -- End of central directory record
-    zh:write(numberToByteString(0x06054b50, 4)) -- signature
-    zh:write(numberToByteString(0, 2)) -- number of this disk
-    zh:write(numberToByteString(0, 2)) -- number of disk with start of central directory
-    zh:write(numberToByteString(#self.files, 2)) -- total number of entries in the central dir on this disk
-    zh:write(numberToByteString(#self.files, 2)) -- total number of entries in the central dir
-    zh:write(numberToByteString(size_of_central_directory, 4))
-    zh:write(numberToByteString(central_directory_offset, 4))
-    zh:write(numberToByteString(0, 2)) -- zip file comment length
-    zh:close()
+    libarchive.archive_write_close(self.archive)
+    libarchive.archive_write_free(self.archive)
+    self.archive = nil
     return true
 end
 
 -- Open zipfile
 function ZipWriter:open(zipfilepath)
-    self.files = {}
-    self.in_open_file = false
-    -- set modification date and time of files to now
-    local t = os.date("*t")
-    self.started_date = bit.bor(
-        bit.lshift(t.year-1980, 9),
-        bit.lshift(t.month,     5),
-        bit.lshift(t.day,       0)
-    )
-    self.started_time = bit.bor(
-        bit.lshift(t.hour,     11),
-        bit.lshift(t.min,       5),
-        bit.rshift(t.sec+2,     1)
-    )
-    self.ziphandle = io.open(zipfilepath, "wb")
-    if not self.ziphandle then
-        return nil
+    self.archive = libarchive.archive_write_new()
+    self.started_time = os.time()
+    libarchive.archive_write_set_format_zip(self.archive)
+    if libarchive.archive_write_open_filename(self.archive, zipfilepath) ~= libarchive.ARCHIVE_OK then
+        local err = ffi.string(libarchive.archive_error_string(self.archive))
+        print('archive_write_open_filename failed:', err)
```

Comment:
Nit. I don't really care about single vs double quotes, but we do use double quotes everywhere. :-)